crates/dht/src/time.rs

```diff
@@ -0,0 +1,1051 @@
+        }
+
+        Ok(partials)
+    }
+}
+
+/// Computes what duration is required for a series of time slices.
+///
+/// The duration is computed as the sum of the powers of two from 0 to the factor, multiplied by
+/// the [UNIT_TIME].
+///
+/// Panics if the factor is greater than 53.
+/// Note that the maximum factor changes if the [UNIT_TIME] is changed.
+fn residual_duration_for_factor(factor: u8) -> Duration {
+    if factor >= 54 {
+        panic!("Factor must be less than 54");
+    }
+
+    let mut sum = 1;
+    for i in 1..=factor {
+        sum |= 1u64 << i;
+    }
+
+    Duration::from_secs(sum * UNIT_TIME.as_secs())
+}
+
+/// Combine a series of op hashes into a single hash.
+///
+/// Requires that the op hashes are already ordered.
+/// If the input is empty, then the output is an empty byte array.
+fn combine_op_hashes(hashes: Vec<OpId>) -> bytes::Bytes {
+    let mut out = if let Some(first) = hashes.first() {
+        bytes::BytesMut::zeroed(first.0.len())
+    } else {
+        // `Bytes::new` does not allocate, so if there was no input, then return an empty
+        // byte array without allocating.
+        return bytes::Bytes::new();
+    };
+
+    let iter = hashes.into_iter().map(|x| x.0 .0);
+    for hash in iter {
+        for (out_byte, hash_byte) in out.iter_mut().zip(hash.iter()) {
+            *out_byte ^= hash_byte;
+        }
+    }
+
+    out.freeze()
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use kitsune2_api::{MetaOp, OpStore};
+    use kitsune2_memory::Kitsune2MemoryOpStore;
+    use std::sync::Arc;
+
+    #[test]
+    fn residual() {
+        assert_eq!(UNIT_TIME, residual_duration_for_factor(0));
+        assert_eq!((2 + 1) * UNIT_TIME, residual_duration_for_factor(1));
+        assert_eq!((4 + 2 + 1) * UNIT_TIME, residual_duration_for_factor(2));
+        assert_eq!(
+            (32 + 16 + 8 + 4 + 2 + 1) * UNIT_TIME,
+            residual_duration_for_factor(5)
+        );
+
+        let mask = 0b1111111111000000u16;
+        assert_eq!(
+            Duration::from_secs(!((mask as u64) << 48) * UNIT_TIME.as_secs()),
+            residual_duration_for_factor(53)
+        );
+    }
+
+    #[should_panic(expected = "Factor must be less than 54")]
+    #[test]
+    fn max_residual() {
+        residual_duration_for_factor(54);
+    }
+
+    #[tokio::test]
+    async fn new() {
+        let origin_timestamp = Timestamp::now();
+        let factor = 4;
+        let pt = PartitionedTime::new(origin_timestamp, factor);
+
+        // Full slices would have size 2^4 = 16, so we should reserve space for at least one
+        // of each smaller slice size
+        assert_eq!((8 + 4 + 2 + 1) * UNIT_TIME, pt.min_recent_time);
+    }
+
+    #[tokio::test]
+    async fn from_store() {
+        let origin_timestamp = Timestamp::now();
+        let factor = 4;
+        let store = Arc::new(Kitsune2MemoryOpStore::default());
+        let pt = PartitionedTime::from_store(origin_timestamp, factor, store)
```

Comment:
I was thinking that this is an app level setting but that it affects the space hash. I.e. changing the factor is not something that can be done within a network and attempting to do so would require a new app install. That also means we couldn't change the default without it being a breaking change.

What do you think? Is it okay to prevent this at a higher level than here?
I'm not really sure how to fit it into the store and it feels redundant to store it with every slice hash because it should be a constant.

Reply:
Sure thing, yes. I see your reasoning for this.